### ISSUE TITLE
Usage of init_score results in no boosting

### ISSUE THREAD

**Issue by SebastianBr** [microsoft/LightGBM#4148]
It seems if lightgbm.train is used with an initial score (init_score) it cannot boost this score. Here is a simple example:

    params = {"learning_rate": 0.1,"metric": "binary_logloss","objective": "binary", 
              "boosting_type": "gbdt","num_iterations": 5, "num_leaves": 2 ** 2,
              "max_depth": 2, "num_threads": 1, "verbose": 0}

    x = pd.DataFrame([[1, 0.1, 0.3], [1, 0.1, 0.3], [1, 0.1, 0.3],  
                      [0, 0.9, 0.3], [0, 0.9, 0.3], [0, 0.9, 0.3]], columns=["a", "b", "prob"])
    y = pd.Series([0, 1, 0, 0, 1, 0])

    d_train = lgb.Dataset(x, label=y)
    model = lgb.train(params, d_train)
    y_pred_default = model.predict(x, raw_score=False)
In the case above, no init_score is used. The predictions are correct: y_pred_default = [0.33333333, ... ,0.33333333]

    d_train = lgb.Dataset(x, label=y, init_score=scipy.special.logit(x["prob"]))
    model = lgb.train(params, d_train)
    y_pred_raw = model.predict(x, raw_score=True)
In this part, we assume colum "prob" from x to be our initial guess (maybe by some other model). We apply logit and use it as initial score. However, the model cannot improve and the boosting will always return 0: y_pred_raw = [0, 0, 0, 0, 0, 0]

    y_pred_raw_with_init = scipy.special.logit(x["prob"]) + y_pred_raw

    y_pred = scipy.special.expit(y_pred_raw_with_init)
This part above shows the way I suppose is correct to translate the initial scores together with the boosting back to probabilities. Since the boosting is zero y_pred yields [0.3, ..., 0.3] which is our initial probability.

I am doing it wrong or assuming init_score to do something else here?

**Comment by jmoralez** [comment]
I believe the problem is with your example. You give the model the same samples (x) labeled both as positive (y = 1) and negative (y = 0), so the model can't learn anything from it, which is printed out as a warning:
`[LightGBM] [Warning] No further splits with positive gain, best gain: -inf`.

If you inspect the model (for example using `model.trees_to_dataframe()`) you'll see that it didn't make any splits, so with both datasets it's using the `init_score` as the predictions. In the first case, since you don't explicitly specify `init_score` it uses the logit of the mean of the target (`logit(0.333)`). In the second case it uses the one you specified (`logit(0.3)`).

**Comment by SebastianBr** [comment]
Thanks for your reply. I agree that the example is a bit strange. And you are correct, the first model simply returns the mean of the target and did not create a tree since it doesn't improve the prediction.

However, I assumed that the second model should at least create a tree. It could simply select the first feature "a=0.5" as a split. Then it would also predict a raw score of logit(0.333) (depending on the learning rate of course), which results in lower log loss than a logit of 0.3. Don't you agreee?

So for me it seems as if the model is not trying to improve the provided init_score. From my understanding init_score can be treated as the result of some other initial prediction output and gradient boosting will try to boost this prediction by further reducing the loss. But this is not what I see here.

Or is this some wrong assumption of me?

**Comment by jmoralez** [comment]
> However, I assumed that the second model should at least create a tree. It could simply select the first feature "a=0.5" as a split. Then it would also predict a raw score of logit(0.333) (depending on the learning rate of course), which results in lower log loss than a logit of 0.3. Don't you agreee?

This is `a` vs `y`:
![orig_data](https://user-images.githubusercontent.com/8473587/113482363-b5534480-945b-11eb-9e44-f59eacfa8595.png)

So as you can see a split at `a=0.5` doesn't produce any gain, since you'd still have the same points at y=0 and y=1, therefore lightgbm doesn't use that feature for splitting, same for `b`.

> From my understanding init_score can be treated as the result of some other initial prediction output and gradient boosting will try to boost this prediction by further reducing the loss. But this is not what I see here.

This is correct.

Here's an example:
```python
import lightgbm as lgb
import matplotlib.pyplot as plt
import numpy as np

X = np.arange(4).reshape(-1, 1)
y = np.array([0, 1, 1, 1])
```
This is the data
![example_data](https://user-images.githubusercontent.com/8473587/113482695-0b74b780-945d-11eb-90c4-85f3461eaf00.png)

LightGBM will use the mean of the target as init_score, so here it'd be 0.75, let's try setting that manually:
```python
ds = lgb.Dataset(X, y, init_score=np.array([0.75] * 4))
model = lgb.train({'num_round': 1, 'num_leaves': 2, 'min_data_in_leaf': 1, 'objective': 'l2', 'learning_rate': 1}, ds)
model.predict(X, raw_score=True)
# array([-0.75,  0.25,  0.25,  0.25])
```
So here the boosting subtracts 0.75 for the first sample and adds 0.25 for the others, the final predictions here are 0, 1, 1, 1.

If we try a different init_score:
```python
ds = lgb.Dataset(X, y, init_score=np.array([0.5] * 4))
model = lgb.train({'num_round': 1, 'num_leaves': 2, 'min_data_in_leaf': 1, 'objective': 'l2', 'learning_rate': 1}, ds)
model.predict(X, raw_score=True)
# array([-0.5,  0.5,  0.5,  0.5])
```
As you can see here the boosting is starting from 0.5 in this case, and the tree is either subtracting or adding 0.5 to get to the same final predictions (0, 1, 1, 1).

**Comment by SebastianBr** [comment]
Thanks the examples. I think I finally figured what my problem is. It might be my misunderstanding or a wrong expectation:

I use gradient boosting for probability prediction. Hence my goal is to achieve best likelihood estimates. This is also why I came up with this strange input matrix with duplicates.
So it terms of classification, simply deciding whether an input example belongs to class 1 or 0, the classifier does actually the correct thing. I, however, assumed that it would continue splitting the data until the probability estimates are best and hence, also the log loss would be minimal. So I think it does not have anything to do with the init_score but simply my understanding of the optimization itself.

Here again an example to show what I mean:
```
        import lightgbm as lgb
        import numpy as np

        X = np.array([1, 3, 2, 2]).reshape(-1, 1)
        y = np.array([0, 1, 0, 1])

        ds = lgb.Dataset(X, y)
        model = lgb.train({'num_round': 3 'num_leaves': 2, 'min_data_in_leaf': 1, 'objective': 'l1', 'learning_rate': 1}, ds)
        model.predict(X, raw_score=True)
        # array([0.5, 1.0, 0.5, 0.5])
```

So basically the algorithm stops after one iteration, predicting f(the feature)=3 to be positive, all other three examples to be negative. The L1 loss would be 0.5+0.5 + 0.5 = 1.5 Creating another tree/leaf that would put example with f=1 in it, would result in the prediction vector [0.0, 1.0, 0.5, 0.5] and hence would reduce the loss to 0.5 + 0.5 = 1.0, which would be ideal in my application.
However, it terms of binary prediction, it wouldn't matter since still all three example with f <=2 would end up in the negative class. 

Is this assumption correct? Because in my very first example where I said a=0.5 would be a good split, it would also not change the class assignment (I assume that is what you wanted to show me with your plot). However, creating the split would improve the loss (predictions would change from 0.3 to 0.333).

And if this is correct, and tree splitting would stop before the metric is minimal, is there way to force such a behaviour?

**Comment by jmoralez** [comment]
> Creating another tree/leaf that would put example with f=1 in it, would result in the prediction vector [0.0, 1.0, 0.5, 0.5] and hence would reduce the loss to 0.5 + 0.5 = 1.0, which would be ideal in my application.

In this new example this is actually possible, I believe you're not getting it because of the bins. As you may know, in order to reduce the number of possible splits that LightGBM can perform, the features are first binned and then the possible splits are the bin boundaries. The [default min_data_in_bin is 3](https://lightgbm.readthedocs.io/en/latest/Parameters.html#min_data_in_bin), so it's possible that the split you're referring to (feature <=1) wasn't actually computed here. You can get the result you described in the example reducing that parameter (notice that I set `num_leaves=3` to get that extra split. I also set `objective='l2'` because I'm not sure how the predictions are computed with `l1`):

```python
import lightgbm as lgb
import numpy as np

X = np.array([1, 3, 2, 2]).reshape(-1, 1)
y = np.array([0, 1, 0, 1])

ds = lgb.Dataset(X, y, params={'min_data_in_bin': 1})
model = lgb.train({'num_round': 3, 'num_leaves': 3, 'min_data_in_leaf': 1, 'objective': 'l2', 'learning_rate': 1}, ds)
np.round(model.predict(X), 2)
# array([0. , 1. , 0.5, 0.5])
```
If you change `objective='binary'` you can get roughly the same for classification:

```
ds = lgb.Dataset(X, y, params={'min_data_in_bin': 1})
model = lgb.train({'num_round': 3, 'num_leaves': 3, 'min_data_in_leaf': 1, 'objective': 'binary', 'learning_rate': 1}, ds)
np.round(model.predict(X), 2)
# array([0.02, 0.96, 0.52, 0.52])
```

> And if this is correct, and tree splitting would stop before the metric is minimal, is there way to force such a behaviour?

Boosting will run for as many iterations as you specify and it can reach a very small loss. However, it requires that the features you provide allow it to get there. In this example for the last two samples there's nothing a tree can do to improve, because for f=2, y = 1 and y = 0, so there's no way to use this feature and improve the loss. If you had an extra feature that actually split these samples, then LightGBM would find it and split it to get to a perfect score.

**Comment by SebastianBr** [comment]
You are correct. Your changes lead to a perfect fit. Thanks for this!

However, if the algorithm is finding the loss minimum, I don't understand why in my initial example a split at e.g. a = 0.5 would not do anything. The first three examples will be in one leaf and the other in the second leaf. Of course the mean of y will not change. Is this what you referring at as gain? The logit does not change. Is this the reason the split is not made?

This split would of course only make sense if the initial score is off, then the log loss would improve:

```
        import lightgbm as lgb
        import numpy as np
        import sklearn

        X = pd.DataFrame([[1, 0.1, 0.3], [1, 0.1, 0.3], [1, 0.1, 0.3],
                                         [0, 0.9, 0.3], [0, 0.9, 0.3], [0, 0.9, 0.3]], columns=["a", "b", "prob"])
        y = pd.Series([0, 0, 1, 0, 1, 0])

        ds = lgb.Dataset(X, y, init_score=X["prob"])
        model = lgb.train({'num_round': 3, 'num_leaves': 4, "metric": "logloss", 'min_data_in_bin': 1, 'min_data_in_leaf': 1,
                                        'objective': 'binary', 'learning_rate': 1}, ds)

        raw_y = scipy.special.logit(X["prob"]) + model.predict(X, raw_score=True)
        loss = sklearn.metrics.log_loss(y, scipy.special.expit(raw_y))
        # 0.6391075640678002

        loss = sklearn.metrics.log_loss(y,  np.array([0.333] * 6))
        # 0.6365141707948961
```

Basically the model learns nothing more than the correct offset / intercept in this case. 0.3 is already close but splitting the examples and predicting even a tiny bit more than 0.3 will improve the log loss.

Anyways I think this an extreme case and you proofed that the init_score is used in the correct way, so I assume its not buggy.

**Comment by jmoralez** [comment]
> I don't understand why in my initial example a split at e.g. a = 0.5 would not do anything. The first three examples will be in one leaf and the other in the second leaf. Of course the mean of y will not change. Is this what you referring at as gain? The logit does not change. Is this the reason the split is not made?

A split will only be made if it makes a positive gain (the predictions are better than before). In that example you have 6 samples (2 positive and 4 negative, therefore that leaf has 2/6 = 33% positives), if you make a split at `a <= 0.5` you'll have 3 samples in each leaf (1 positive and 2 negatives or 1/3 = 33% positives), so each leaf is pretty much the same as its parent (which means that performing that split results in no gain and therefore it won't be made). The same applies for `b`.

If you had a feature that could actually split the classes, something like: `X['c'] = [0.2, 0.3, 0.7, 0.4, 0.8, 0.1]`. Then when `X['c'] <= 0.5` `y = 0` and otherwise `y = 1`. This split would separate the classes effectively (have a positive gain) so it would be performed and you could achieve a very small loss.

**Comment by SebastianBr** [comment]
Yes, you are correct and I think we might start going in circles :-)

Let's settle with the fact that a split will result in the same mean of y, so the same positive ratio in each leaf. By looking at this metric alone, there is no gain in splitting. This is exactly as you mentioned before. I got this and totally agree.

But suppose your init_score is off by far, say we predicted y to be only 0.01. When splitting, we should not compare the positive rate of 1/3 in the leafs with the positive rate in the original data of 1/3 but with the "predicted" initial positive rate of 0.01.  So a split would do nothing more than to trigger the model to correct this initial score.

We can simulate this by using a bagging since then the model would, by chance, see a subset of the data that needs correction.  Again an example (note I also switched to l2 norm, since you mentioned this is more familiar to you):
       
```
        import lightgbm as lgb
        import numpy as np
        import sklearn

        X = pd.DataFrame([[1, 0.1, 0.01], [1, 0.1, 0.01], [1, 0.1, 0.01],
                                         [0, 0.9, 0.01], [0, 0.9, 0.01], [0, 0.9, 0.01]], columns=["a", "b", "prob"])
        y = pd.Series([0, 0, 1, 0, 1, 0])

        ds = lgb.Dataset(X, y, init_score=X["prob"])
        model = lgb.train({'num_round': 10, 'num_leaves': 4, "metric": "logloss", 'min_data_in_bin': 1, 'min_data_in_leaf': 1,
                                       'objective': 'mse', 'learning_rate': 1, 'bagging_freq': 1, 'bagging_fraction': 0.7}, ds)
        loss = sklearn.metrics.log_loss(y, X["prob"] + model.predict(X, raw_score=True))
        # bagging_fraction=1.0 => 1.541756952565031
        # bagging_fraction=0.7 => 0.6365141682948128

        optimal_loss = sklearn.metrics.log_loss(y, np.array([1/3] * 6))
        # 0.6365141682948127
```
I understood that this is not how the algorithm works. But maybe you see my point that there is something to gain from a split, but the algorithms 'gain' is still zero.
